tests/util/tests_Regex.cpp

```diff
@@ -0,0 +1,76 @@
+#define DOCTEST_CONFIG_IMPLEMENT_WITH_MAIN
+
+#include <doctest/doctest.h>
+#include <rdf4cpp/rdf.hpp>
+
+using namespace rdf4cpp::rdf::regex;
+
+TEST_SUITE("regex") {
+    TEST_CASE("replacement translation") {
+        Regex const r{"[0-9]"};
+        RegexReplacer repl_dummy;
+
+        SUBCASE("no escape") {
+            CHECK_THROWS(repl_dummy = r.make_replacer("$"));
+
+            {
+                std::string s = "Hello 99 World";
+                r.make_replacer(R"($0_)").regex_replace(s);
+                CHECK(s == "Hello 9_9_ World");
+            }
+        }
+
+        SUBCASE("1 escape") {
+            {
+                std::string s = "Hello 99 World";
+                r.make_replacer(R"(\$)").regex_replace(s);
+                CHECK(s == "Hello $$ World");
+            }
+
+            {
+                std::string s = "Hello 99 World";
+                r.make_replacer(R"(\$0)").regex_replace(s);
+                CHECK(s == "Hello $0$0 World");
+            }
+        }
+
+        SUBCASE("2 escape") {
+            CHECK_THROWS(repl_dummy = r.make_replacer(R"(\\$)"));
+
+            {
+                std::string s = "Hello 99 World";
+                r.make_replacer(R"(\\$0)").regex_replace(s);
+                CHECK(s == R"(Hello \9\9 World)");
+            }
+        }
+
+        SUBCASE("3 escape") {
+            {
+                std::string s = "Hello 99 World";
+                r.make_replacer(R"(\\\$)").regex_replace(s);
+                CHECK(s == R"(Hello \$\$ World)");
+            }
+
+            {
+                std::string s = "Hello 99 World";
+                r.make_replacer(R"(\\\$0)").regex_replace(s);
+                CHECK(s == R"(Hello \$0\$0 World)");
+            }
+        }
+
+        SUBCASE("4 escape") {
+            CHECK_THROWS(repl_dummy = r.make_replacer(R"(\\\\$)"));
+
+            {
+                std::string s = "Hello 99 World";
+                r.make_replacer(R"(\\\\$0)").regex_replace(s);
+                CHECK(s == R"(Hello \\9\\9 World)");
+            }
+        }
+
+        SUBCASE("stray backslash") {
+            // TODO
+            // CHECK_THROWS(r.make_replacer(R"(Hello \ World)"));
```

Comment:
Also something for an issue?

Reply:
This shouldn't work because it's an invalid escape (even in RE2 syntax as far as I can tell).
RE2 just accepts it and does some weird things here. I probably need to change the translation somehow to catch these cases manually.

Reply:
I.e. with regex r = `[0-9]` this passes:

```
std::string s = "Hello 1 World";
r.make_replacer(R"(Hello \ World)").regex_replace(s);
CHECK(s == "Hello Hello  World");
```

Reply:
I would guess it is an escaped space. Like in Latex `\ `.

Reply:
Do you think the behaviour is fine this way or is it important that we catch that?

Reply:
I just noticed something. Everything after a `\` which is not part of a reference to the match in the replacement string just gets deleted. You can see it in the example I gave above and also:

```
std::string s = "0";
r.make_replacer(R"(Hello \hgfhf World)").regex_replace(s);
CHECK(s == "Hello ");
```

It doesn't seem to matter what comes after the backslash if it isnt another backslash or a number the rest of the replacement just gets ignored. Why

Reply:
That is very odd. I would say we have for now some preliminary regex support. We state that the syntax follows google RE2. So users are warned. I would first complete features and then come back to this kind of fine-tuning.

Reply:
#141